### ISSUE TITLE
ENH overlap-add convolutions

### ISSUE THREAD

**Issue by eickenberg** [kymatio/kymatio#305]
I think this is standard practice for 1D signals, because they often have to deal with very long signals wrt filter size, but this can be applied across the board: Imagine a very big 2D image, or, more realistically, a 3D array which is actually a video of 2D frames which you want to analyze with spatiotemporal gabors. This will require overlap-add in the time axis.

So the general setup might be something where you specify the axes that you want to have overlap-add in, and it could be a `WaveletTransform` or `OverlapAddConvolution` module that takes care of it.

**Comment by lostanlen** [comment]
Yes. This is something that is available in scattering.m for example and it's really useful

**Comment by janden** [comment]
The other potential advantage here, from what I understand, is that it may give us a way to compute filters independent of image size, since we're storing them in the spatial/time domain.

**Comment by eickenberg** [comment]
Working with 1D data of shape `(384, 300000)` is showing me the limitations of not having this implemented. My GPU maxes out in memory at `(30, 300000)`, which, if we made this into one long line, would correspond to like 400s of mono audio at 22050. It might be nice to be able to go a bit above that without having to manually chunk data etc

**Comment by lostanlen** [comment]
As a bioacoustician, i routinely work on input waveforms that are a billion samples long or more. Overlap-add is essential in this case. And 👍 for working independently of image size. This opens the path for training fully convolutional models and recurrent models

**Comment by MuawizChaudhary** [comment]
Maybe one of you can explain the basic concept, and I can try to implement this one of these days?

**Comment by lostanlen** [comment]
[copied from a Discord discussion]

N>>T is the typical use case of variable-length input. 

To address it, pick a buffer size M>T  (M is the maximal support of the wavelets/phi). M depends solely on the Scattering1D object, not on the input itself.

Then, given a signal x of length N>>M, divide x into C overlapping chunks of length M. I think that a default overlap of 50% is sensible.
Note that padding is only needed for the first and last chunk. For all the chunks in between, we can use overlap as a form of "natural padding".

In scattering.m i have experimented with two variants of this chunking procedure: "Hann" and "Tukey", after the name of the window functions applied to every chunk. In both variants, the number of scattering frames per chunk should be a multiple of 4. (so N=4T sounds about right?)

Here's the layout i imagine: 
```
chunk A: 0 1 2 3
chunk B:     0 1 2 3
chunk C:         0 1 2 3
chunk D:             0 1 2 3
```

(again, this example takes 4 frames per chunk but the idea can work with more frames too)

In the Hann variant, one would sum the responses on a per-chunk basis (no FFT needed!) and end up with the sequence [A, B, C, D] 
The appeal of this is that it potentially allows us to skip the phi  conv stage and replace it by this per-chunk sum which has linear complexity.
So i think that the Hann variant could be the best choice if (1)average=True  and (2) the averaging scaleT is greater or equal than the maximum wavelet scale
Now, in the Tukey variant, one would discard the 25% left portion and the 25% right portion of every chunk
```
chunk A:   1 2 
chunk B:       1 2 
chunk C:           1 2 
chunk D:               1 2 
```
 
and then stitch the segments together:
[A1, A2, B1, B2, C1, C2, D1, D2]

In my opinion the Tukey variant is the best choice if average=False or if T  is smaller than the maximum wavelet scale

@OverLordGoldDragon your comments are welcome!
